### DISCUSSION TITLE
posts/025-astro-workers/

### DISCUSSION THREAD

**Discussion by giscus** [Yamabuki-bakery/tasty-discussion#11] [Announcements]
# posts/025-astro-workers/

俺想在博客裏面建朋友圈，但是就是不想自建服務器！💢👎🏻

http://localhost:4321/posts/025-astro-workers/

**Comment by PermeationLoop** [comment]
家人們誰董阿🍉😄

**Comment by NagatoCute** [comment]
你ao

**Comment by PermeationLoop** [comment]
你好，這裏是 2024 年 12 月 12 日，我成功把博客升級到了 Astro 5.0；新功能是沒體驗到一點，但是把代碼裏面的 slug 全部改爲 id 花了俺兩天時間💢😅

**Comment by PermeationLoop** [comment]
昨晚做了一個夢，夢到這個博客內嵌的朋友圈系統被人抄襲了，設計佈局和代碼實現全被抄走了，，，
俺不知道是該高興還是該難過😔

**Reply by PermeationLoop** [reply]
現在是 2026 年，美夢（噩夢）成真了，真的被全部抄走了，具體是哪位請看我的友鏈 😭

**Comment by SaltedFishesNG** [comment]
其实还有一个复古的思路，不借助任何第三方的东西，直接写HTML和CSS，并添加少量JS作为锦上添花。
(然后我就缺乏设计思路了(saltedfishes.com)，不知道homepage和nav怎么样设计才好。qwq)

依赖其他框架生成的页面总是会存在些许不稳定因素(例如tailwindcss竟然不少地方依赖js)
而且用了一些CSS lib会使你的HTML部分扎堆出现`<div class="" id="">`的情况，而那些CSS选择器简直就像是完全没见过HTML长什么样似的。
太不优雅了（确信

（好怪，这个站点的dark theme居然是完全由js控制的，但其实
`@media (prefers-color-scheme: dark)`是好的，也可以用js控制，例如:<https://github.com/hackademix/noscript-web/blob/main/js/darklight.js>）
(哦对，还有就是非必要其实不建议内联CSS或JS。)

**Reply by PermeationLoop** [reply]
妳的復古思路很好，許多 web 愛好者都會通過手撕 HTML 來精簡自己的網站，如果妳有興趣，可以看看俺以前寫的《[👉🏻100KB HTML 手撕大賽](https://100kb.pages.dev/)》頁面。

不過這個頁面只是個半成品，俺沒能完成他，原因是越寫到後面，頁面上的要素越多，就越難以管理，我也逐漸開始希望能有個 XX 框架來幫我封裝一下可以復用的組件，幫我自動生成 CSS。實際上這就是第三方框架出現的原因，特別是像博客這樣有一定規模的網站，如果我要對所有的 HTML 親力親為，我哪裏還有時間去寫文章，，，

俺拜訪了妳的網站（並且通過友情連結認出了妳是 Kun 群的群友），妳的網站很有特色，我覺得如果妳想要繼續設計下去的話，可以試試走「ASCII ART 風格」的设计语言。例如

```
🏠[[Meow~]]🏠           [🥰About me] [📦Blog Archive] [🔗Link Exchange] [✅TODO] 

----------

This is SaltedFishes's Site...

（其他內容）

```

沒有必要排斥框架，比如說 Astro，他能又快又好地幫你生成想要的佈局，讓妳能專心在內容寫作上面。你提到了 Tailwind 依賴 JS 的問題，在最新版本裏面他已經全面改用 CSS 來寫配置文件了，不會有 JS 來摻一腳了，只不過我還沒升級。把 佈局寫在 class 裏面已經差不多是慣例了，因爲我也試過 module.css 文件到處亂飛的寫法，還是寫在 class 裏面再編譯會比較乾淨一點。

我也試過 @media 來控制夜間模式，可是俺總是在編寫手動切換的時候遇到問題，所以採用了現在這種簡單起見的方法。精力有限，沒法面面俱到，主打一個能跑就行。說不定以後有空才會改進（

**Reply by SaltedFishesNG** [reply]
qwq，其实你这种颇具视觉冲击力的设计风格我是学不来的(( 然后走的极简风格路线。
我的站点主要问题其实是/Archive/和/pages/风格差异不小，其次才是homepage"过时了"(因Link Exchange独立)。

(顺 便 一 提 ,几乎所有ASCII ART是不用emoji的，当时没有emoji。并且它只有两种颜色)

黑暗模式的话还有一个较新的CSS用法<https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/light-dark>

CSS选择器如果减少对class和id的依赖可以减少10%~30%多余的框架代码。

例如这个`<nav class="astro-3ef6ksr2" id="nav-menu">`，整个HTML页面nav只出现过一次，直接用元素定位更好；然后这个独特元素下面的子元素也是独特的，也应该用元素选择器。
如果你这么写的话大概就能完全理解我说的"CSS就像是没见过HTML长什么样似"的这句话
像这种独特的元素，HTML有不少呢，避免什么东西都用div和span才是优雅(确信

---

当CSS足够优秀的时候，\<main\>下的内容甚至可以不使用任何class或id。
新文章只需要复制上一篇文章的头和尾，改一下标题和简介就行。最后再用一个脚本或手动处理下Archives和Tgas部分。

当然，用例如 www.getzola.org 或ejs这种直接套模板也是一种优雅解决方法。

**Reply by PermeationLoop** [reply]
謝謝妳的讚賞，在開始這個博客之前，俺一直都在用圓角的設計風格，這個 blog 是一次換口味的嘗試。希望妳也能找到讓妳滿意的設計。

妳的站點 /Archive/ 和 /pages/ 風格有差異的原因是不是因爲妳縫合了兩個 hexo 主題？俺看了一下文章的數量，覺得手動遷移主題並不難，妳有空可以試一下。

俺說的 ASCII ART 只是一個舉例，在製作妳的新主題的時候，並不是要照搬所謂的「習慣」，而是形成妳自己的設計風格。比如說，爲了易讀性往純文本裏面加 emoji 🥰，或者用巧妙的方法插入圖片來增加視覺效果。我覺得，網站做出來，讓人來讀，覺得好讀，易讀，才是最大的意義。

TailwindCSS 其實並不只是一個「把 style 寫進 class 裏」的投機取巧框架，他的上限其實很高，他能夠讓妳定義自己的 spacing、color、breakpoint，甚至 typography，從而構建自己的設計系統；而 utility class 的好處是快速構建，需要復用的時候再用 @apply 或 component 抽象重複部分。

沒有必要去糾結 class 或者 id 裏面的代碼量，因爲開發人員的體驗也是體驗。妳提到的 <nav> 元素，還有許多你沒提到的其他元素，裏面有很多 id 和 class 其實是寫給我自己看的，為的就是能讓我之後快速看懂我過去的想法。

CSS 不需要知道 HTML 長什麼樣。在 Tailwind 模式下，我們用 class 來修飾 HTML，一個 class 只幹一件事情，HTML 和 CSS 互相不需要知道對方在幹甚麼，這對於維護有極大好處。CSS 不需要太優秀，他只需要做好自己的那點事，不要干涉我的其他代碼就行了。

------

> 新文章只需要复制上一篇文章的头和尾，改一下标题和简介就行。最后再用一个脚本或手动处理下Archives和Tgas部分。

我以前做過一個項目，就是重構這麼一個純手撕的網站。以往，管理員每發佈一個新網頁，都需要手動複製頭尾，然後手動修改變化的部分。這導致一個問題，當需要改設計，比如說要改 header 和 footer 的時候，就需要修改幾十個文件，讓整個網站沒有 scale up 的能力。後面俺使用 Next.js 的組件化寫法重做了網站，從那以後管理員發新網頁就只需要寫點 md 了，開發效率少說提升 5 倍吧，更重要的是節省了不計其數的找 bug 修 bug 的時間。

俺希望妳能考慮並取捨一下框架帶來的好處，，，

**Reply by SaltedFishesNG** [reply]
其实/pages/是我自己写的(
markdown的话，各个实现都有不小差距，写着写着就成HTML了()然后慢慢发展成现在直接写HTML了。

scale up确实是很大的问题，只能确保你之前操作足够规范，用IDE的功能批量搜索替换就可。

**Comment by PermeationLoop** [comment]
費了九牛二虎之力，俺昨天終於把本博升級到了 TailwindCSS V4。
模板的原作者的代碼真的是一坨大的，升級難度極高，這應該是歷史上最重量的一次升級了。

步驟：
1. 照着 V3 的配置文件 tailwind.config.cjs 的內容，在 base.css 裏面重寫 @theme 的顏色和字體配置；
1. 刪除 V3 配置文件，在 base.css 裏面 `@import "tailwindcss"` 和 `@plugin "@tailwindcss/typography"`
1. pnpm 刪除 tailwindcss 和 @astrojs/tailwind
1. 運行 pnpx astro add tailwind 進行一個重裝
1. 由於 v4 在寫 @apply 的時候，必須要在旁邊 `@reference 'tailwindcss'`，因此老版本那種到處亂拉屎 apply 的寫法就會報錯。俺尋思，你到處亂七八糟 apply 的寫法，和手撕 css 有什麼區別？Tailwind 就應該寫在 class 裏面
   1. 我讓 Copilot 幫我批量搜索，然後把那些 apply 的規則全部移動到 class 裏面
   1. 手動修復了一些壞掉的排版
   1. 做完這一步以後，網站才終於能正常運行不報錯 😭
1. 用手機版瀏覽器和 Safari 檢查佈局，修復壞掉的部分

TailwindCSS 官方提供的升級工具在遇到這種屎山代碼的時候根本跑不起來，給的報錯也莫名其妙，真的是無從下手。
手動升級前後大概花費 4 個多小時，再加上之前嘗試升級的無數次失敗，可能償還技術債就是這種體驗吧 😭

也算是了卻一樁心事了，，，

**Comment by PermeationLoop** [comment]
### 隨想：Blog 下雲從計劃到破產 😇

我很想和 Cloudflare 說再見，這不僅是由於 Cloudflare 導致網速不增反降，也是因爲我想充分利用那些吃灰已久的服務器。所以我老早就在盤算着如何給博客下雲。

要下雲，就要把下面這幾個關鍵部件搬出 Cloudflare：

- CI/CD
- Static Assets (Cloudflare Pages)
- API Endpoints (Cloudflare Workers)
- Observability

最先下雲的是 CICD，我在 GitHub Actions 手撕了一條編譯流水線，實現了「Built on GitHub, Hosted on CF」。我想着，如果之後要 self-host 整個網站，改改這條流水線應該不難。

接着我用 Grafana 搭建了自己的 Observability 系統。爲了收集訪問數據，俺騰出了一臺閒置服務器，在上面用 docker compose 起了一個 caddy 來反代整個網站，還順手獲得了加速效果。

然後我磨刀霍霍指向 Workers。這是 Cloudflare 的 SaaS，按道理來說應該是不可能下雲的…… 但是 CF 真是菩薩啊，開源了他們的 workers 實現 (workerd)，我直接一通手撕，把它塞進了 Docker，於是 Workers 也下雲了。😁

最後就只剩下靜態網頁的部分了。乍一看這應該是最簡單的部分，CICD build 完直接塞進 Caddy 就完事…… 可惜，事情沒那麼簡單。首先我需要安全地把 GitHub Actions 編譯出的產物導入服務器，然後掛載到 Caddy 裏面。

光是這一點就非常複雜了，爲了安全性，我沒法在服務器上開放 SSH 和 Webhook 端點以供 CICD 上傳文件。就算能開放，我還是要做一大堆設置用戶權限，編寫部署腳本，設計 asset 的版本管理…… 等非常複雜的工作。關鍵是，這些東西難以全部塞進 docker，也沒法自動化。換一臺服務器，我就要把這些東西重新再設置一遍。工作量都快趕上手撕一套 Kubenetes 了。

爲了避免做出原子彈打蚊子的行爲，我決定就此收手，髒活累活交給 Cloudflare，然後車欠。😇

---

雖然下雲計劃沒能撐到最後，但是部分下雲也不是一無是處，美國服務器帶來了網速的提升，Grafana 讓我能夠一窺互聯網的晴雨，自建 workerd 讓我不必在意 Workers 免費額度的限制，這些都是實打實的好處 🥰

**Reply by SaltedFishesNG** [reply]
其实你不一定非要依赖GitHub Actions，你可以直接把git push代码到服务器上，然后在服务器上触发自动脚本编译和它。
这样你连GitHub都不需要依赖了

**Reply by PermeationLoop** [reply]
這確實是一種方法。但是最大的問題是，我沒有拋棄 GitHub 的理由啊，我的代碼和配置文件都在上面。
而且把代碼放在一臺面向公網的公開服務器上不是甚麼好做法。。

**Reply by SaltedFishesNG** [reply]
那就让服务器定期从GitHub clone代码

**Reply by PermeationLoop** [reply]
代碼不應該出現在公開服務器上，而且，定期 clone 屬於「解決一個問題但造成三個新問題」的做法。

**Comment by PermeationLoop** [comment]
😇 **近期博客下雲進度**

俺已經解決了 CICD 推送靜態文件的問題，並已經平穩運行了幾個月。具體做法就是在服務器上用 Python 起了一個小 HTTP 服務器，用來接收 CICD 推送過來的靜態文件，解壓給 caddy、workerd 並重啓他們。

GitHub Action 那邊則在 Astro 編譯完成後打包靜態文件，用 curl 上傳到服務器上。整個過程通過 HTTPS 傳送和 API key 驗證確保安全性，Python 的部分也待在 Docker 裏面，力圖減少攻擊面。這是俺目前能想到的最實用穩定的方法了。

另外，最近俺從 Grafana 裏面發現有一些腳本小子在通過全球的 botnet 爬我的文章。雖然只造成了 0 點傷害，但是卻污染了統計數據和日誌。因此俺準備要導入 CrowdSec 的 IP 黑名單和入侵檢測系統 😈 萬一有無辜群友被誤封的話，俺提前在這裏誠懇地向妳道歉，，，
